libflux/src/semantic/nodes.rs

```diff
@@ -0,0 +1,567 @@
+pub struct FunctionExpression<'a> {
+    pub loc: Location<'a>,
+    pub type_info: TypeInformation,
+
+    pub params: Vec<FunctionParameter<'a>>,
+    pub body: FunctionBody<'a>,
+}
+
+impl<'a> FunctionExpression<'a> {
+    pub fn pipe(&self) -> Option<&FunctionParameter<'a>> {
+        for p in &self.params {
+            if p.is_pipe {
+                return Some(p);
+            }
+        }
+        None
+    }
+
+    pub fn defaults(&self) -> Vec<&FunctionParameter<'a>> {
+        let mut ds = Vec::new();
+        for p in &self.params {
+            match p.default {
+                Some(_) => ds.push(p),
+                None => (),
+            }
+        }
+        ds
+    }
+}
+
+// FunctionParameter represents a function parameter.
+#[derive(Debug, PartialEq, Clone)]
+pub struct FunctionParameter<'a> {
+    pub loc: Location<'a>,
+
+    pub is_pipe: bool,
+    pub key: Identifier<'a>,
+    pub default: Option<Expression<'a>>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct BinaryExpression<'a> {
+    pub loc: Location<'a>,
+    pub type_info: TypeInformation,
+
+    pub operator: ast::OperatorKind,
+    pub left: Expression<'a>,
+    pub right: Expression<'a>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct CallExpression<'a> {
+    pub loc: Location<'a>,
+    pub type_info: TypeInformation,
+
+    pub callee: Expression<'a>,
+    pub arguments: ObjectExpression<'a>,
+    pub pipe: Option<Expression<'a>>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct ConditionalExpression<'a> {
+    pub loc: Location<'a>,
+    pub type_info: TypeInformation,
+
+    pub test: Expression<'a>,
+    pub consequent: Expression<'a>,
+    pub alternate: Expression<'a>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct LogicalExpression<'a> {
+    pub loc: Location<'a>,
+    pub type_info: TypeInformation,
+
+    pub operator: ast::LogicalOperatorKind,
+    pub left: Expression<'a>,
+    pub right: Expression<'a>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct MemberExpression<'a> {
+    pub loc: Location<'a>,
+    pub type_info: TypeInformation,
+
+    pub object: Expression<'a>,
+    pub property: String,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct IndexExpression<'a> {
+    pub loc: Location<'a>,
+    pub type_info: TypeInformation,
+
+    pub array: Expression<'a>,
+    pub index: Expression<'a>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct ObjectExpression<'a> {
+    pub loc: Location<'a>,
+    pub type_info: TypeInformation,
+
+    pub with: Option<IdentifierExpression<'a>>,
+    pub properties: Vec<Property<'a>>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct UnaryExpression<'a> {
+    pub loc: Location<'a>,
+    pub type_info: TypeInformation,
+
+    pub operator: ast::OperatorKind,
+    pub argument: Expression<'a>,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct Property<'a> {
+    pub loc: Location<'a>,
+
+    pub key: PropertyKey<'a>,
+    pub value: Option<Expression<'a>>,
```

Comment:
Why is this optional?

Reply:
There can be cases in which the objects do not specify values. For example:

```
{a, b}
```

Is equivalent to

```
{a: a, b: b}
```

Now I see your point. I guess I could remove the  optional expression and assign an IdentifierExpr in that case.

Reply:
I have done that in the last commit.